### PR TITLE
[desktop] Add Gitpod to Settings sync

### PR DESCRIPTION
**Pull request by filiptronicek**
Add Gitpod as an authentication and Settings Sync provider via the `gitpod.gitpod-desktop` extension.

- [x] invalidate bad token(s) on startup (https://github.com/gitpod-io/openvscode-server/pull/139/commits/27f4c9c4f15155d369692c1c2617a16e828f20a6)
- [x] fix the `openExternal` issue
- [x] implement code exchange
- [ ] Fix 401 WS Error
- [ ] contribute Gitpod endpoint config to the manifest

Blocked by https://github.com/gitpod-io/gitpod/pull/6272

### CODE REVIEW

**Comment by akosyakov** [comment]
@filiptronicek Could you squash in one commit and rebase on `gp-code/main`? I tried to but did not figure out what is the first commit and which changes should be applied during rebase.

File: extensions/gitpod/src/registerAuth.ts [hunk 1]
```diff
@@ -0,0 +1,388 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+
+/// <reference path='../../../src/vs/vscode.d.ts'/>
```
Comment (akosyakov) [inline extensions/gitpod/src/registerAuth.ts:5]: let's rename this file to `auth.ts`

**Review by akosyakov — COMMENTED** [review]
It looks promising, but the code structure needs more love. Also I could not find how we invalidate existing tokens. It would be nice to align in these regards with GH auth extension.

File: extensions/gitpod/src/registerAuth.ts [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+	const onDidChangeSessionsEmitter = new vscode.EventEmitter<vscode.AuthenticationProviderAuthenticationSessionsChangeEvent>();
+	logger('Registering authentication provider...');
+	context.subscriptions.push(vscode.authentication.registerAuthenticationProvider('gitpod', 'Gitpod', {
+		onDidChangeSessions: onDidChangeSessionsEmitter.event,
+		getSessions: async (scopes: string[]) => {
+			const sessions: vscode.AuthenticationSession[] = [];
+			if (!scopes) {
+				return Promise.resolve(sessions);
+			}
+			const storedSessionsJSON = await context.secrets.get('gitpod.authSessions') || '[]';
```
[the remark this answers is not in the document]
Comment (filiptronicek) [reply extensions/gitpod/src/registerAuth.ts:373]: Not yet, 😬, adding it to the PR tasklist.

**Comment by iamtangram** [comment]
ddd

**Comment by jeanp413** [comment]
continuation in https://github.com/gitpod-io/openvscode-server/pull/337

**Pull request closed without merging**